pkg/agent/controller/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,658 @@
+			CaptureLength: len(packet.Data()),
+			Length:        len(packet.Data()),
+		}
+		err = captureState.pcapngWriter.WritePacket(ci, packet.Data())
+		if err != nil {
+			return fmt.Errorf("couldn't write packet: %w", err)
+		}
+		klog.V(2).InfoS("capture packet", "name", captureState.name, "count",
+			captureState.numCapturedPackets, "packet", packet.String())
+
+		reachTarget := captureState.numCapturedPackets == captureState.maxNumCapturedPackets
+		// use rate limiter to reduce the times we need to update status.
+		if reachTarget || captureState.updateRateLimiter.Allow() {
+			pc, err := c.packetCaptureLister.Get(captureState.name)
+			if err != nil {
+				return fmt.Errorf("get PacketCapture failed: %w", err)
+			}
+			// if reach the target. flush the file and upload it.
+			if reachTarget {
+				if err := captureState.pcapngWriter.Flush(); err != nil {
+					return err
+				}
+				if err := c.uploadPackets(pc, captureState.pcapngFile); err != nil {
+					return err
+				}
+				if err := captureState.pcapngFile.Close(); err != nil {
+					return err
+				}
+				if err := c.setPacketsFilePathStatus(pc.Name); err != nil {
+					return err
+				}
+			}
+			err = c.updatePacketCaptureStatus(pc, crdv1alpha1.PacketCaptureRunning, "", captureState.numCapturedPackets)
+			if err != nil {
+				return fmt.Errorf("failed to update the PacketCapture: %w", err)
+			}
+			klog.InfoS("Updated PacketCapture", "PacketCapture", klog.KObj(pc), "numCapturedPackets", captureState.numCapturedPackets)
+		}
+	}
+	return nil
+}
+
+func (c *Controller) getTargetPodIP(podRef *crdv1alpha1.PodReference, isIPv6 bool) (net.IP, error) {
+	srcPodInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podRef.Name, podRef.Namespace)
+	var result net.IP
+	if len(srcPodInterfaces) > 0 {
+		if isIPv6 {
+			result = srcPodInterfaces[0].GetIPv6Addr()
+		} else {
+			result = srcPodInterfaces[0].GetIPv4Addr()
+		}
+	} else {
+		pod, err := c.kubeClient.CoreV1().Pods(podRef.Namespace).Get(context.TODO(), podRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get pod %s/%s: %v", podRef.Namespace, podRef.Name, err)
+		}
+		podIPs := make([]net.IP, len(pod.Status.PodIPs))
+		for i, ip := range pod.Status.PodIPs {
+			podIPs[i] = net.ParseIP(ip.IP)
+		}
+		if isIPv6 {
+			ip, err := util.GetIPWithFamily(podIPs, util.FamilyIPv6)
+			if err != nil {
+				return nil, err
+			} else {
+				result = ip
+			}
+		} else {
+			result = util.GetIPv4Addr(podIPs)
+		}
+	}
+	if result == nil {
+		family := "IPv4"
+		if isIPv6 {
+			family = "IPv6"
+		}
+		return nil, errors.New(fmt.Sprintf("cannot find IP with %s AddressFamily for pod %s/%s", family, podRef.Namespace, podRef.Name))
+	}
+	return result, nil
+}
+
+func (c *Controller) createMatchPacket(pc *crdv1alpha1.PacketCapture) (*binding.Packet, error) {
+	packet := new(binding.Packet)
+	if pc.Spec.Packet == nil {
+		pc.Spec.Packet = &crdv1alpha1.Packet{
+			IPFamily: v1.IPv4Protocol,
```

Comment:
Not sure how Traceflow handle the default protocol, I suppose we should check if the cluster is ipv4/ipv6 or a dual stack cluster first via networkConfig? `networkConfig.IPv4Enabled /  networkConfig.IPv6Enabled`